### DISCUSSION TITLE
DX: Getting the Current URL in Server Components

### DISCUSSION THREAD

**Discussion by landsman** [vercel/next.js#65385] [Polls]
Do you think the existence of an article like this: https://www.propelauth.com/post/getting-url-in-next-server-components, or [issues like this](https://github.com/vercel/next.js/issues/43704) shows us that the developer experience is bad in this area?

**Comment by crawlchange** [comment]
The feature requests talking about this are readily closed by NextJS maintainers, who say that the behavior is intentional because the layout doesn't re-render for different pages:

https://github.com/vercel/next.js/issues/47108

What is, then, their excuse for templates? How come templates do not have access to the current route?

**Comment by leerob** [comment]
This article is inaccurate and I would highly recommend not following the patterns it is showing. Futher, I would recommend to the authors to remove this utility from their package, as it is likely to break.

Please see this comment for the correct answer: https://github.com/vercel/next.js/issues/43704#issuecomment-2090798307

**Reply by andrew-propelauth** [reply]
Hey @leerob! I actually wrote that article and I’ve also read that comment.

Ignoring the client component pieces, which I hopefully captured in the post, if I understand the comment correctly, it basically amounts to “in a server component you can only get the dynamic parts of the route, but you should know the static parts to put together the full path/URL.”

Could you explain why the middleware approach will break? I’ve seen it both in a lot of the comments as well in a few other repos. We’re happy to remove it if it’s not the right approach, but right now it seems like the best of mostly not great options.

**Reply by landsman** [reply]
See. This weird complexity is my question about DX. I don't want to work like this.

**Reply by slorber** [reply]
@leerob I'm also unsure to understand how this pattern would break

This @h3rmanj comment remains unanswered: https://github.com/vercel/next.js/issues/43704#issuecomment-2096045800

> Like last time, the response still doesn't answer why `headers()` and `cookies()` exist, while `pathname()` does not. If I opt into dynamic rendering for my layout, what is so different about the pathname, from headers or cookies? All the reasons you list for pathname not being accessible, seems like they would apply to headers and cookies too.

Answering this would probably help us all understand the problem.

**Reply by BrendonSled** [reply]
The middleware approach may not work when navigating routes. This was my findings the last time I used next.

Edit: I just verified this approach still does not work on the latest version of next. Navigating routes will not rerender components that use the headers.

Edit 2: I should also add this is when using the headers under a layout. There may be other footguns which may be why this approach isn't recommended.

**Comment by crawlchange** [comment]
@leerob Still no explanation why TEMPLATES cannot access the path of the current page being rendered under them.

**Comment by armouti** [comment]
The DX of next has definitely dropped from 9/10 about 2-3 years ago, to 3/10 at best now

**Comment by MangoMarcus** [comment]
> @leerob Still no explanation why TEMPLATES cannot access the path of the current page being rendered under them.

Is there any clarity now on why page templates (as opposed to layout templates) cannot access the current path name @leerob ?

I also don't understand why it's not possible given that `usePathname()` _does_ work when rendered statically on the server at build time

Is it because server components are used/cached across different routes regardless of if they're in a page or a layout template?
